File: DataStructures/doublyLinkedLists/dll.py
class Node:
    def __init__(self, val):
        self.val = val
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def __str__(self):
        node = self.head
        str = "[ "
        while node:
            str += f"{node.val}, "
            node = node.next
        return str + "]"

    def push(self, val):
        newNode = Node(val)
        if not self.head:
            self.head = newNode
            self.tail = newNode
        else:
            self.tail.next = newNode
            newNode.prev = self.tail
            self.tail = newNode
        self.length += 1
        return self

    def pop(self):
        if not self.tail:
            return None

        poppedNode = self.tail
        print('poppedNode Val', self.tail)

        if self.tail == 1:
            self.tail = None
            self.head = None
        else:
            self.tail = poppedNode.prev
            poppedNode.prev = None
            self.tail.next = None

        self.length -= 1
        return poppedNode

    def shift(self):
        if not self.head:
            return None

        poppedNode = self.head

        if self.tail == 1:
            self.head = None
            self.tail = None
        else:
            self.head = poppedNode.next
            self.head.prev = None
            poppedNode.next = None

        self.length -= 1
        return poppedNode

    def unshift(self, val):
        new_node = Node(val)

        if not self.head:
            self.head = new_node
            self.tail = new_node
        else:
            self.head.prev = new_node
            new_node.next = self.head
            self.head = new_node

        self.length += 1
        return self

    def get(self, index):
        if index < 0 or index >= self.length:
            return None

        if index < (self.length/2):
            counter = 0
            node = self.head
            while counter < index:
                node = node.next
                counter += 1
        else:
            counter = self.length-1
            node = self.tail
            while counter > index:
                node = node.prev
                counter -= 1
        return node

    def set(self, index, val):
        node = self.get(index)
        if node:
            node.val = val
            return True
        return False

    def insert(self, index, val):
        if index < 0 or index > self.length:
            return None

        new_node = Node(val)

        if self.length == index:
            self.push(val)
        elif index == 0:
            self.unshift(val)
        else:
            current = self.get(index)
            before = current.prev
            before.next = new_node
            new_node.prev = before
            new_node.next = current
            current.prev = new_node
        self.length += 1
        return self

    def remove(self, index):
        if index < 0 or index >= self.length or self.length == 0:
            return None
        if self.length-1 == index:
            self.pop()
        elif index == 0:
            self.shift()
        else:
            removedNode = self.get(index)
            before = removedNode.prev
            after = removedNode.next
            before.next = after
            after.prev = before
            removedNode.prev = None
            removedNode.next = None
        self.length -= 1
        return removedNode


dll = DoublyLinkedList()
dll.push(2)
dll.push(4)
dll.push(6)
dll.push(8)
dll.push(10)
dll2 = DoublyLinkedList()

print(dll.head.val)
print(dll.tail.val)
print(dll.length)

print('popped Node', dll.pop().val)
print('popped length', dll.length)
print('popped tail', dll.tail.val)

print('shifted head node val', dll.shift().val)

print('put the head back on', dll.unshift(2))
print('new length should be 4', dll.length)
print('dll.__str__', dll.__str__())
print('node at index 2', dll.get(2).val)
print('node at index 1', dll.get(1).val)
print('setting node at idx 1 to 9', dll.set(1, 9))
print('inserting node at idx 1 to 12', dll.insert(1, 12))
print('removing node at idx 2', dll.remove(2).val)
print('dll.__str__', dll.__str__())

File: DataStructures/doublyLinkedLists/dll.js
class Node {
  constructor(val) {
    this.val = val;
    this.next = null;
    this.prev = null;
  }
}

class DoublyLinkedList {
  constructor() {
    this.head = null;
    this.tail = null;
    this.length = 0;
  }
  push(val) {
    newNode = new Node(val)
    if (!this.head) {
      this.head = newNode
      this.tail = newNode
    }
    else {

      newNode.prev = this.tail
      this.tail.next = newNode
      this.tail = newNode

    }
    this.length++
    return this


  }

  pop() {
    if (!this.head) return undefined;
    var poppedNode = this.tail;
    if (this.length === 1) {
      this.head = null;
      this.tail = null;
    } else {
      this.tail = poppedNode.prev;
      this.tail.next = null;
      poppedNode.prev = null;
    }
    this.length--;
    return poppedNode;
  }
  shift() {
    if (!this.head) return undefined;
    var oldHead = this.head;
    if (this.length == 1) {
      this.head = null;
      this.tail = null;
    } else {
      this.head = oldHead.next;
      this.head.prev = null;
      oldHead.next = null;
    }
    this.length--;
    return oldHead;
  }

  unshift(val) {
    var newNode = new Node(val);
    if (this.length === 0) {
      this.head = newNode;
      this.tail = newNode;
    } else {
      this.head.prev = newNode;
      newNode.next = this.head;
      this.head = newNode;
    }
    this.length++;
    return this;
  }

  get(index) {
    if (index < 0 || index >= this.length) return null;

    if (index <= this.length / 2) {
      console.log("WORKING FROM START");
      var count = 0;
      var current = this.head;

      while (count !== index) {
        current = current.next; //traverse the list
        count++;
      }
    } else {
      console.log("WORKING FROM END");
      var count = this.length - 1;
      var current = this.tail;

      while (count !== index) {
        current = current.prev;
        count--;
      }
    }
    return current;
  }

  set(index, val) {
    var node = this.get(index);
    if (node != null) {
      node.val = val;
      return true;
    }

    return false;
  }

  insert(index, val) {
    if (index < 0 || index > this.length) return false;
    if (index == 0) return !!this.unshift(val);
    if (index === this.length) return !!this.push(val);

    var beforeNode = this.get(index - 1);
    var newNode = new Node(val);
    var afterNode = beforeNode.next;

    if (beforeNode != null) {
      beforeNode.next = newNode;
      newNode.next = afterNode;
      afterNode.prev = newNode;
      newNode.prev = beforeNode;
      this.length++;
    }
    return true;
  }

  remove(index) {
    if (index < 0 || index > this.length) return false;
    if (the index == 0, shift)



  }
}

var list = new DoublyLinkedList();
list.shift();

list.push("hermione");
list.push("harry");
list.push("ron");
console.log(list);

list.unshift("snape");
list.unshift("malfoy");
list.unshift("dumbledore");

console.log(list);
console.log("item 4, ", list.get(4));
// console.log("item 1 set, ", list.set(1, "HAGRIDDDD"));
console.log("item 1 insert, ", list.insert(1, "TONKS"));
console.log(list);
